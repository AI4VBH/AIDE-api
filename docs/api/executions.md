# Executions

## `GET /executions/:workflow_instance_id/tasks/:execution_id/artifacts`

This returns the task output artifacts.

### Parameters

| Parameter Type | Label | Description |
|----------------|-------|-------------|
| URL | `:workflow_instance_id` | The workflow instance ID |
| URL | `:execution_id` | The task execution ID |

### Responses

This endpoint returns a dictionary of filename and MINIO object key.

| Code | Description |
|------|-------------|
| `200` | Returns output artifacts |
| `404` | Workflow instance not found or task execution not found. |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
curl --location --request GET 'http://localhost:5000/executions/9df2f059-7669-4404-b02c-427510d7db04/tasks/c2fa42a0-4a96-46b0-b599-8467ca6d3f7e/artifacts'
```

### Example Response

```json
{
  "dicom-report": "9df2f059-7669-4404-b02c-427510d7db04/c2fa42a0-4a96-46b0-b599-8467ca6d3f7e/dicom-report"
}
```

---

## `GET /executions/artifact-download?key=:object_key`

This redirects the request to a pre-signed URL from MINIO enabling the downloading of file.

### Parameters

| Parameter Type | Label | Description |
|----------------|-------|-------------|
| Query | `:object_key` | The object key for the artifact you are attempting to download |

### Responses

| Code | Description |
|------|-------------|
| `302` | Redirect to the pre-signed MINIO url |
| `400` | Missing `:object_key` |
| `500` | Internal server error |

### Example Request

```bash
curl --location --request GET 'http://localhost:5000/executions/artifact-download?key=9df2f059-7669-4404-b02c-427510d7db04/c2fa42a0-4a96-46b0-b599-8467ca6d3f7e/dicom-report'
```

---

## `GET /executions/:workflow_instance_id/tasks/:execution_id/metadata`

Returns the task execution metadata.

### Response

Returns a dictionary.

| Code | Description |
|------|-------------|
| `200` | Returns execution metadata |
| `404` | Workflow instance not found or task execution not found. |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
curl --location --request GET 'http://localhost:5000/executions/9df2f059-7669-4404-b02c-427510d7db04/tasks/c2fa42a0-4a96-46b0-b599-8467ca6d3f7e/metadata'
```

### Example Response

```json
{
  "metadata-1": "value",
  "metadata-2": "value"
}
```

