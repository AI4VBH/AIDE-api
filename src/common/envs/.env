PORT=5000

KEYCLOAK_URL='http://0.0.0.0:8443'
KEYCLOAK_MAIN_REALM='master'
KEYCLOAK_REALM='Demo-Realm'
KEYCLOAK_CLIENTID='nest-app'
KEYCLOAK_SECRET='83790b4f-48cd-4b6c-ac60-451a918be4b9'
KEYCLOAK_STATIC_ROLES='["admin", "clinician"]'

DATABASE_HOST='localhost'
DATABASE_NAME='nest_api'
DATABASE_USER='postgres'
DATABASE_PASSWORD='admin'
DATABASE_PORT=5432

ELASTIC_HOST=''
ELASTIC_PORT=''
ELASTIC_USE_SSL=''
ELASTIC_USERNAME=''
ELASTIC_PASSWORD=''
ELASTIC_INDEX=''

MONAI_API_HOST='https://localhost:8081/'
MIG_API_HOST='http://localhost:5010/'

MINIO_HOST='localhost'
MINIO_PORT=9001
MINIO_USE_SSL=false
MINIO_BUCKET='monaideploy'
MINIO_ACCESS_KEY='wOzmbchico52AMp3'
MINIO_SECRET_KEY='zyoT9SiMLiTCvNT0DS636FMiN4yi2258'
