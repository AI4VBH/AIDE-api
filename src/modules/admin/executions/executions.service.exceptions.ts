export enum ExecutionsServiceExceptionType {
  MISSING_TASK = 1,
}

export class ExecutionsServiceException extends Error {
  constructor(type: ExecutionsServiceExceptionType, message: string) {
    super(message);

    this.type = type;
  }

  readonly type: ExecutionsServiceExceptionType;
}
