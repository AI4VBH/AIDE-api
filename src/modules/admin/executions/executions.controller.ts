import {
  BadRequestException,
  Controller,
  Get,
  Inject,
  Logger,
  Param,
  ParseUUIDPipe,
  Query,
  Redirect,
  UseFilters,
} from '@nestjs/common';
import { Public } from 'nest-keycloak-connect';
import ExternalServerExceptionFilter from 'shared/http/external-server-exception.filter';
import { ExecutionsService } from './executions.service';

@Controller('executions')
@UseFilters(ExternalServerExceptionFilter)
export class ExecutionsController {
  @Inject(ExecutionsService)
  private readonly executionsService: ExecutionsService;

  @Get(':workflow_instance_id/tasks/:execution_id/artifacts')
  async getWorkflowInstanceArtifacts(
    @Param('workflow_instance_id') workflowInstanceId: string,
    @Param('execution_id', ParseUUIDPipe) executionId: string,
  ) {
    return this.executionsService.getWorkflowInstanceArtifacts(
      workflowInstanceId,
      executionId,
    );
  }

  @Get('artifact-download')
  @Public()
  @Redirect()
  async getArtifactDownloadUrl(@Query('key') file: string) {
    const logger = new Logger('ArtifactDownload');

    if (!file || !file.trim()) {
      throw new BadRequestException('key query value is missing');
    }

    logger.log(`Generating pre-signed url for key: '${file}'`);
    const url = await this.executionsService.getArtifactUrl(file);
    logger.log(`Pre-signed URL generated: ${url}`);

    return { url };
  }

  @Get(':workflow_instance_id/tasks/:execution_id/metadata')
  getWorkflowInstanceMetadata(
    @Param('workflow_instance_id') workflowInstanceId: string,
    @Param('execution_id', ParseUUIDPipe) executionId: string,
  ) {
    return this.executionsService.getWorkflowInstanceMetadata(
      workflowInstanceId,
      executionId,
    );
  }
}
