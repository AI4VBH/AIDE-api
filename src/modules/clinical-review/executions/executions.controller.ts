import { Controller, Get } from '@nestjs/common';
import { Executions, ExecutionsSearchResult } from './executions.interface';
import { ExecutionsService } from './executions.service';

@Controller('executions')
export class ExecutionsController {
  constructor(private readonly executionsService: ExecutionsService) {}

  @Get()
  async getModels() {
    return await this.executionsService.getExecutions();
  }
}
