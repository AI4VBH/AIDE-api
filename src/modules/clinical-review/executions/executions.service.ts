import { Injectable } from '@nestjs/common';
import { Executions, ExecutionsSearchResult } from './executions.interface';

@Injectable()
export class ExecutionsService {
  async getExecutions() {
    // console.log('searching');
    // console.log(searchResult);
    return {};
  }
}
