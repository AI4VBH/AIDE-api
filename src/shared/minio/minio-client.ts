import { Inject, Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { Client, ItemBucketMetadata } from 'minio';
import { parseBoolean } from 'shared/util/parseBoolean';
import { Readable } from 'stream';

@Injectable()
export class MinioClient extends Client {
  constructor(@Inject(ConfigService) config: ConfigService) {
    super({
      endPoint: config.get<string>('MINIO_HOST'),
      port: parseInt(config.get('MINIO_PORT')),
      useSSL: parseBoolean(config.get('MINIO_USE_SSL')),
      accessKey: config.get<string>('MINIO_ACCESS_KEY'),
      secretKey: config.get<string>('MINIO_SECRET_KEY'),
    });

    this.bucketName = config.get<string>('MINIO_BUCKET');

    const scheme = parseBoolean(config.get('MINIO_USE_SSL')) ? 'https' : 'http';

    this.baseUrl = `${scheme}://${config.get<string>('MINIO_HOST')}:${parseInt(
      config.get('MINIO_PORT'),
    )}`;
  }

  private readonly bucketName: string;
  public readonly baseUrl: string;

  async getPresignedObjectUrl(objectName: string): Promise<string> {
    return await this.presignedGetObject(this.bucketName, objectName, 60 * 60);
  }

  async getObjectByName(objectName: string): Promise<Readable> {
    return await this.getObject(this.bucketName, objectName);
  }

  async getObjectMetadata(objectName: string): Promise<ItemBucketMetadata> {
    const { metaData } = await this.statObject(this.bucketName, objectName);
    return metaData;
  }
}

export class MinoiClientException extends Error {
  constructor(message: string, stack?: string) {
    super(message);

    this.stack = stack;
  }
}
