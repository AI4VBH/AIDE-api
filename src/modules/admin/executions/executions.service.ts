import { HttpService } from '@nestjs/axios';
import { Inject, Injectable } from '@nestjs/common';
import { lastValueFrom } from 'rxjs';
import { MinioClient, MinoiClientException } from 'shared/minio/minio-client';
import internal from 'stream';
import { MonaiWorkflowInstance } from '../workflowinstances/workflowinstances.interface';
import {
  ExecutionsServiceException,
  ExecutionsServiceExceptionType,
} from './executions.service.exceptions';

@Injectable()
export class ExecutionsService {
  @Inject(HttpService)
  private readonly httpService: HttpService;

  @Inject(MinioClient)
  private readonly minioClient: MinioClient;

  public get minioBaseUrl(): string {
    return this.minioClient.baseUrl;
  }

  async getWorkflowInstanceArtifacts(
    workflowInstanceId: string,
    executionId: string,
  ) {
    const response = await lastValueFrom(
      this.httpService.get<MonaiWorkflowInstance>(
        `/workflowinstances/${workflowInstanceId}`,
      ),
    );

    const task = (response.data?.tasks ?? []).find(
      (t) => t.execution_id === executionId,
    );

    if (!task) {
      throw new ExecutionsServiceException(
        ExecutionsServiceExceptionType.MISSING_TASK,
        'Task not found',
      );
    }

    const artifacts = task?.output_artifacts ?? {};

    return artifacts;
  }

  async getArtifactUrl(file: string) {
    try {
      const url = await this.minioClient.getPresignedObjectUrl(file);
      return url;
    } catch (exception) {
      throw new MinoiClientException(exception.message);
    }
  }

  async getArtifact(
    file: string,
  ): Promise<{ contentType: string; stream: internal.Readable }> {
    try {
      const stream = await this.minioClient.getObjectByName(file);
      const metadata = await this.minioClient.getObjectMetadata(file);

      return { stream, contentType: metadata['content-type'] };
    } catch (exception) {
      throw new MinoiClientException(exception.code);
    }
  }

  async getWorkflowInstanceMetadata(
    workflowInstanceId: string,
    executionId: string,
  ) {
    const response = await lastValueFrom(
      this.httpService.get<MonaiWorkflowInstance>(
        `/workflowinstances/${workflowInstanceId}`,
      ),
    );

    const task = (response.data?.tasks ?? []).find(
      (t) => t.execution_id === executionId,
    );

    if (!task) {
      throw new ExecutionsServiceException(
        ExecutionsServiceExceptionType.MISSING_TASK,
        'Task not found',
      );
    }

    return task?.result ?? {};
  }
}
