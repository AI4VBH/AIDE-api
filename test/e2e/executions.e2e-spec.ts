/*
 * Copyright 2022 Guy’s and St Thomas’ NHS Foundation Trust
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';
import { HttpModule } from '@nestjs/axios';
import { ConfigModule } from '@nestjs/config';
import * as request from 'supertest';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { HttpConfigService } from 'shared/http/http.service';
import ExecutionsMock from 'test/test_data/mocks/executions/executions-index';
import { MinioClient } from 'shared/minio/minio-client';
import { ExecutionsController } from 'modules/admin/executions/executions.controller';
import { ExecutionsService } from 'modules/admin/executions/executions.service';

const server = setupServer();
const testMonaiBasePath = 'https://localhost:7337';

describe('/executions Integration Tests', () => {
  let app: INestApplication;

  beforeAll(() => {
    server.listen({
      onUnhandledRequest: 'bypass',
    });
    server.printHandlers();
  });

  afterAll(() => {
    server.close();
  });

  beforeEach(async () => {
    server.resetHandlers();

    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [
        ConfigModule.forRoot({
          isGlobal: true,
          load: [
            () => ({
              MONAI_API_HOST: testMonaiBasePath,
              MINIO_HOST: 'localhost',
              MINIO_PORT: 9000,
              MINIO_USE_SSL: false,
              MINIO_EXTERNAL_URL: 'http://some-external-minio.site/document',
              MINIO_ACCESS_KEY: 'access-key',
              MINIO_SECRET_KEY: 'secret-key',
              MINIO_BUCKET: 'bucket-name',
            }),
          ],
        }),
        HttpModule.registerAsync({
          useClass: HttpConfigService,
        }),
      ],
      controllers: [ExecutionsController],
      providers: [ExecutionsService, MinioClient],
    }).compile();
    app = moduleFixture.createNestApplication();
    await app.init();
  });

  it('(GET) /executions/:workflow_inst_id/tasks/:execution_id/artifacts no artifacts', async () => {
    server.use(
      rest.get(
        `${testMonaiBasePath}/workflowinstances/73909439-57da-465e-b529-2addd36cbcd9`,
        (_req, res, ctx) => {
          return res(ctx.json(ExecutionsMock.noArtifacts));
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      '/executions/73909439-57da-465e-b529-2addd36cbcd9/tasks/fd89296e-29b7-4d54-be27-26e185db820c/artifacts',
    );
    expect(response.body).toMatchObject({});
    expect(response.status).toBe(200);
  });

  it('(GET) /executions/:workflow_inst_id/tasks/:execution_id/artifacts with artifacts', async () => {
    server.use(
      rest.get(
        `${testMonaiBasePath}/workflowinstances/a67a7af7-068b-44b8-a81b-def7b3e5403b`,
        (_req, res, ctx) => {
          return res(ctx.json(ExecutionsMock.withArtifacts));
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      '/executions/a67a7af7-068b-44b8-a81b-def7b3e5403b/tasks/3b9d94b9-4285-45d4-bea9-491fa62b8f91/artifacts',
    );
    expect(response.body).toMatchSnapshot();
    expect(response.status).toBe(200);
  });

  it('(GET) /executions/:workflow_inst_id/tasks/:execution_id/artifacts workflow instance does not exist', async () => {
    server.use(
      rest.get(
        `${testMonaiBasePath}/workflowinstances/41c5778c-9957-4cfd-be7a-c0bbff5c7cca`,
        (_req, res, ctx) => {
          return res(
            ctx.status(404),
            ctx.json(ExecutionsMock.nonExistentWorkflow),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      '/executions/41c5778c-9957-4cfd-be7a-c0bbff5c7cca/tasks/3b9d94b9-4285-45d4-bea9-491fa62b8f91/artifacts',
    );
    expect(response.body).toMatchSnapshot();
    expect(response.status).toBe(404);
  });

  it('(GET) /executions/:workflow_inst_id/tasks/:execution_id/artifacts execution id does not exist', async () => {
    server.use(
      rest.get(
        `${testMonaiBasePath}/workflowinstances/a67a7af7-068b-44b8-a81b-def7b3e5403b`,
        (_req, res, ctx) => {
          return res(
            ctx.status(200),
            ctx.json(ExecutionsMock.nonExistentExecutionId),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      '/executions/a67a7af7-068b-44b8-a81b-def7b3e5403b/tasks/6b9d94b9-4285-45d4-bea9-491fa62b8f88/artifacts',
    );
    expect(response.body).toMatchSnapshot();
    expect(response.status).toBe(404);
  });

  it('(GET) /executions/:workflow_inst_id/tasks/:execution_id/artifacts invalid workflow instance GUID', async () => {
    server.use(
      rest.get(
        `${testMonaiBasePath}/workflowinstances/invalidguid`,
        (_req, res, ctx) => {
          return res(
            ctx.status(400),
            ctx.json(ExecutionsMock.invalidWorkflowId),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      '/executions/invalidguid/tasks/3b9d94b9-4285-45d4-bea9-491fa62b8f91/artifacts',
    );
    expect(response.body).toMatchSnapshot();
    expect(response.status).toBe(400);
  });

  it('(GET) /executions/:workflow_inst_id/tasks/:execution_id/artifacts invalid execution GUID', async () => {
    const response = await request(app.getHttpServer()).get(
      '/executions/ccfae83b-fc6d-475e-8dc0-6c5ea1904699/tasks/invalidGuid/artifacts',
    );
    expect(response.body).toMatchSnapshot();
    expect(response.status).toBe(400);
  });

  it.each([408, 500, 501, 502, 503, 504])(
    '(GET) /payloads correct status when MONAI gives general error with code %s',
    async (code) => {
      server.use(
        rest.get(
          `${testMonaiBasePath}/workflowinstances/a67a7af7-068b-44b8-a81b-def7b3e5403b`,
          (_req, res, ctx) => {
            return res(ctx.status(code));
          },
        ),
      );
      const response = await request(app.getHttpServer()).get(
        '/executions/a67a7af7-068b-44b8-a81b-def7b3e5403b/tasks/6b9d94b9-4285-45d4-bea9-491fa62b8f88/artifacts',
      );
      expect(response.body).toMatchObject({
        message:
          'An error occurred with an external service (MONAI, Clinical Review)',
        statusCode: 500,
      });
      expect(response.statusCode).toBe(500);
    },
  );

  it('(GET) /executions/artifact-download?key=minio-object-key minio with download link', async () => {
    server.use(
      rest.get('http://localhost:9000/bucket-name', (_, res, ctx) => {
        return res(ctx.status(200));
      }),
      rest.get(
        'http://localhost:9000/bucket-name/minio-object-key',
        (_, res, ctx) => {
          return res(ctx.status(200), ctx.body('hello world'));
        },
      ),
      rest.head(
        'http://localhost:9000/bucket-name/minio-object-key',
        (_, res, ctx) => {
          return res(ctx.status(200));
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      '/executions/artifact-download?key=minio-object-key',
    );

    expect(response.status).toBe(200);
  });

  it('(GET) /executions/artifact-download minio object key does not exist', async () => {
    const response = await request(app.getHttpServer()).get(
      '/executions/artifact-download',
    );
    expect(response.body).toMatchObject({});
    expect(response.status).toBe(400);
  });

  it.each([408, 500, 501, 502, 503, 504])(
    '(GET) /executions/artifact-download?key=minio-object-key correct status when MINIO gives general error with code %s',
    async (code) => {
      server.use(
        rest.get(`http://localhost:9000/bucket-name`, (_req, res, ctx) => {
          return res(ctx.status(code));
        }),
      );
      const response = await request(app.getHttpServer()).get(
        '/executions/artifact-download?key=minio-object-key',
      );
      expect(response.body).toMatchObject({
        message: 'An issue occurred with the MINIO service',
        statusCode: 500,
      });
      expect(response.statusCode).toBe(500);
    },
  );

  it('(GET) /executions/${workflow_instance_id}/tasks/${execution_id}/metadata json object', async () => {
    server.use(
      rest.get(
        `${testMonaiBasePath}/workflowinstances/a67a7af7-068b-44b8-a81b-def7b3e5403b`,
        (req, res, ctx) => {
          return res(ctx.json(ExecutionsMock.metadata));
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      '/executions/a67a7af7-068b-44b8-a81b-def7b3e5403b/tasks/3b9d94b9-4285-45d4-bea9-491fa62b8f91/metadata',
    );
    expect(response.body).toMatchSnapshot();
    expect(response.status).toBe(200);
  });

  it('(GET) /executions/${workflow_instance_id}/tasks/${execution_id}/metadata empty json', async () => {
    server.use(
      rest.get(
        `${testMonaiBasePath}/workflowinstances/a67a7af7-068b-44b8-a81b-def7b3e5403b`,
        (req, res, ctx) => {
          return res(ctx.json(ExecutionsMock.metadataEmpty));
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      '/executions/a67a7af7-068b-44b8-a81b-def7b3e5403b/tasks/3b9d94b9-4285-45d4-bea9-491fa62b8f91/metadata',
    );
    expect(response.body).toMatchObject({});
    expect(response.status).toBe(200);
  });

  it('(GET) /executions/${workflow_instance_id}/tasks/${execution_id}/metadata metadata - workflow instance does not exist', async () => {
    server.use(
      rest.get(
        `${testMonaiBasePath}/workflowinstances/a67a7af7-068b-44b8-a81b-def7b3e5403b`,
        (req, res, ctx) => {
          return res(
            ctx.status(404),
            ctx.json(ExecutionsMock.nonExistentWorkflow),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      '/executions/a67a7af7-068b-44b8-a81b-def7b3e5403b/tasks/3b9d94b9-4285-45d4-bea9-491fa62b8f91/metadata',
    );
    expect(response.body).toMatchSnapshot();
    expect(response.status).toBe(404);
  });

  it('(GET) /executions/${workflow_instance_id}/tasks/${execution_id}/metadata metadata - execution id does not exist', async () => {
    server.use(
      rest.get(
        `${testMonaiBasePath}/workflowinstances/a67a7af7-068b-44b8-a81b-def7b3e5403b`,
        (req, res, ctx) => {
          return res(
            ctx.status(200),
            ctx.json(ExecutionsMock.nonExistentExecutionId),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      '/executions/a67a7af7-068b-44b8-a81b-def7b3e5403b/tasks/6b9d94b9-4285-45d4-bea9-491fa62b8f88/metadata',
    );
    expect(response.body).toMatchSnapshot();
    expect(response.status).toBe(404);
  });

  it('(GET) /executions/${workflow_instance_id}/tasks/${execution_id}/metadata metadata - invalid workflow instance GUID', async () => {
    server.use(
      rest.get(
        `${testMonaiBasePath}/workflowinstances/invalidguid`,
        (_req, res, ctx) => {
          return res(
            ctx.status(400),
            ctx.json(ExecutionsMock.invalidWorkflowId),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      '/executions/invalidguid/tasks/3b9d94b9-4285-45d4-bea9-491fa62b8f91/metadata',
    );
    expect(response.body).toMatchSnapshot();
    expect(response.status).toBe(400);
  });

  it('(GET) /executions/${workflow_instance_id}/tasks/${execution_id}/metadata metadata - invalid execution GUID', async () => {
    const response = await request(app.getHttpServer()).get(
      '/executions/ccfae83b-fc6d-475e-8dc0-6c5ea1904699/tasks/invalidGuid/metadata',
    );
    expect(response.body).toMatchSnapshot();
    expect(response.status).toBe(400);
  });

  it.each([408, 500, 501, 502, 503, 504])(
    '(GET) /executions/${workflow_instance_id}/tasks/${execution_id}/metadata correct status when MONAI gives general error with code %s',
    async (code) => {
      server.use(
        rest.get(
          `${testMonaiBasePath}/workflowinstances/a67a7af7-068b-44b8-a81b-def7b3e5403b`,
          (_req, res, ctx) => {
            return res(ctx.status(code));
          },
        ),
      );
      const response = await request(app.getHttpServer()).get(
        '/executions/a67a7af7-068b-44b8-a81b-def7b3e5403b/tasks/6b9d94b9-4285-45d4-bea9-491fa62b8f88/metadata',
      );
      expect(response.body).toMatchObject({
        message:
          'An error occurred with an external service (MONAI, Clinical Review)',
        statusCode: 500,
      });
      expect(response.statusCode).toBe(500);
    },
  );
});
