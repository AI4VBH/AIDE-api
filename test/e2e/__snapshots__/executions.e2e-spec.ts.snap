// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/executions Integration Tests (GET) /executions/\${workflow_instance_id}/tasks/\${execution_id}/metadata json object 1`] = `
{
  "key": {
    "name": "Some metadata",
    "score": 34,
  },
}
`;

exports[`/executions Integration Tests (GET) /executions/\${workflow_instance_id}/tasks/\${execution_id}/metadata metadata - execution id does not exist 1`] = `
{
  "message": "Task not found",
  "statusCode": 404,
}
`;

exports[`/executions Integration Tests (GET) /executions/\${workflow_instance_id}/tasks/\${execution_id}/metadata metadata - invalid execution GUID 1`] = `
{
  "message": "Validation failed (uuid is expected)",
  "statusCode": 400,
}
`;

exports[`/executions Integration Tests (GET) /executions/\${workflow_instance_id}/tasks/\${execution_id}/metadata metadata - invalid workflow instance GUID 1`] = `
{
  "detail": "Failed to validate workflow instance id, not a valid guid",
  "instance": "/workflowinstances/invalidguid",
  "status": 400,
  "title": "Bad Request",
  "traceId": "00-7d36d37cd21021e68d552618bc5bfc3c-335958bf8d6d9ad0-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.1",
}
`;

exports[`/executions Integration Tests (GET) /executions/\${workflow_instance_id}/tasks/\${execution_id}/metadata metadata - workflow instance does not exist 1`] = `
{
  "detail": "Failed to validate id, workflow instance not found",
  "instance": "/workflows/41c5778c-9957-4cfd-be7a-c0bbff5c7cca",
  "status": 404,
  "title": "Not Found",
  "traceId": "00-1a488bac0cc6ae854a5b6ab0c237dd8f-5ab3204912c13ca3-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.4",
}
`;

exports[`/executions Integration Tests (GET) /executions/:workflow_inst_id/tasks/:execution_id/artifacts execution id does not exist 1`] = `
{
  "message": "Task not found",
  "statusCode": 404,
}
`;

exports[`/executions Integration Tests (GET) /executions/:workflow_inst_id/tasks/:execution_id/artifacts invalid execution GUID 1`] = `
{
  "message": "Validation failed (uuid is expected)",
  "statusCode": 400,
}
`;

exports[`/executions Integration Tests (GET) /executions/:workflow_inst_id/tasks/:execution_id/artifacts invalid workflow instance GUID 1`] = `
{
  "detail": "Failed to validate workflow instance id, not a valid guid",
  "instance": "/workflowinstances/invalidguid",
  "status": 400,
  "title": "Bad Request",
  "traceId": "00-7d36d37cd21021e68d552618bc5bfc3c-335958bf8d6d9ad0-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.1",
}
`;

exports[`/executions Integration Tests (GET) /executions/:workflow_inst_id/tasks/:execution_id/artifacts with artifacts 1`] = `
{
  "report-dicom": "d801bd68-381d-4ff3-95a8-85c46380e075/workflows/a67a7af7-068b-44b8-a81b-def7b3e5403b/3b9d94b9-4285-45d4-bea9-491fa62b8f91/report-dicom",
}
`;

exports[`/executions Integration Tests (GET) /executions/:workflow_inst_id/tasks/:execution_id/artifacts workflow instance does not exist 1`] = `
{
  "detail": "Failed to validate id, workflow instance not found",
  "instance": "/workflows/41c5778c-9957-4cfd-be7a-c0bbff5c7cca",
  "status": 404,
  "title": "Not Found",
  "traceId": "00-1a488bac0cc6ae854a5b6ab0c237dd8f-5ab3204912c13ca3-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.4",
}
`;

exports[`/executions Integration Tests (GET) /executions/artifact-download?key=minio-object-key minio with download link 1`] = `{}`;
